File: Framework/LanguageSupport/cs/MMICSharp/Adapter/ThriftAdapterImplementation.cs
// SPDX-License-Identifier: MIT
// The content of this file has been developed in the context of the MOSIM research project.
// Original author(s): Felix Gaisbauer

using System;
using System.Collections.Generic;
using MMIStandard;
using System.Linq;
using CSharpAdapter;
using MMICSharp.Common;
using System.Diagnostics;

namespace MMICSharp.Adapter
{
    /// <summary>
    /// Implementation of the thrift adapter functionality
    /// </summary>
    [Serializable]
    public class ThriftAdapterImplementation : MMIAdapter.Iface
    {

        /// <summary>
        /// The assigned mmuInstantiator
        /// </summary>
        private readonly IMMUInstantiation mmuInstantiator;

        /// <summary>
        /// Variable for storing the skeleton access
        /// </summary>
        private MSkeletonAccess.Iface skeletonAccess;


        /// <summary>
        /// The assigned session data
        /// </summary>
        private readonly SessionData SessionData;


        /// <summary>
        /// Basic constructor
        /// </summary>
        public ThriftAdapterImplementation(SessionData sessionData, IMMUInstantiation mmuInstantiator)
        {
            this.SessionData = sessionData;
            this.mmuInstantiator = mmuInstantiator;
        }

        #region Adapter specific features

        /// <summary>
        /// Working
        /// </summary>
        /// <returns></returns>
        public virtual Dictionary<string, string> GetStatus()
        {
            Dictionary<string, string> status = new Dictionary<string, string>();

            try
            {
                status = new Dictionary<string, string>
                {
                    { "Running since", SessionData.StartTime.ToLongDateString() + " " + SessionData.StartTime.ToLongTimeString() },
                    { "Total Sessions", SessionData.SessionContents.Count.ToString() },
                    { "Loadable MMUs", SessionData.MMUDescriptions.Count.ToString() }
                };


                try
                {
                    status.Add("Version", System.Reflection.Assembly.GetEntryAssembly().GetName().Version.ToString());

                }
                catch (Exception)
                {
                
                }
                if (SessionData.LastAccess == DateTime.MinValue)
                    status.Add("Last Access", "None");

                else
                    status.Add("Last Access", SessionData.LastAccess.ToLongDateString() + " " + SessionData.LastAccess.ToLongTimeString());
            }
            catch (Exception e)
            {
                status.Add("Exception", e.Message);
            }

            return status;
        }


        /// <summary>
        /// Returns all available MMUs
        /// </summary>
        /// <returns></returns>
        public virtual List<MMUDescription> GetLoadableMMUs()
        {
            return SessionData.MMUDescriptions;
        }


        /// <summary>
        /// Returns the scene contained within the adapter
        /// </summary>
        /// <param name="sessionID"></param>
        /// <returns></returns>
        public virtual List<MSceneObject> GetScene(string sessionID)
        {
            SessionContent sessionContent = null;

            MBoolResponse sessionResult = SessionData.GetSessionContent(sessionID, out sessionContent);

            if (sessionResult.Successful)
            {
                //Set the last access time
                sessionContent.UpdateLastAccessTime();

                Logger.Log(Log_level.L_INFO, "Transfer all scene objects");

                return sessionContent.SceneBuffer.GetSceneObjects();
            }

            else
            {
                Debug.Fail(sessionResult.LogData.ToString());
            }

            return new List<MSceneObject>();
        }


        /// <summary>
        /// Returns the deltas of the last frame
        /// </summary>
        /// <param name="sessionID"></param>
        /// <returns></returns>
        public virtual MSceneUpdate GetSceneChanges(string sessionID)
        {
            SessionContent sessionContent = null;

            MBoolResponse sessionResult = SessionData.GetSessionContent(sessionID, out sessionContent);

            if (sessionResult.Successful)
            {
                //Set the last access time
                sessionContent.UpdateLastAccessTime();

                return sessionContent.SceneBuffer.GetSceneChanges();
            }

            else
            {
                Logger.Log(Log_level.L_ERROR, sessionResult.LogData.ToString());
            }

            return new MSceneUpdate();
        }


        /// <summary>
        /// 
        /// </summary>
        /// <param name="mmus"></param>
        /// <param name="sessionID"></param>
        /// <returns>A mapping from MMUID to a specific instance id</returns>
        public virtual Dictionary<string,string> LoadMMUs(List<string> mmus, string sessionID)
        {
            SessionContent sessionContent = null;
            SessionID idContainer = new SessionID(sessionID);

            //Get the session content for the id
            MBoolResponse sessionResult = SessionData.GetSessionContent(sessionID, out sessionContent);

            //Skip if invalid session result
            if (!sessionResult.Successful)
            {
                Logger.Log(Log_level.L_ERROR, "Cannot generate session content");
                return new Dictionary<string, string>();
            }

            //Set the last access time
            sessionContent.UpdateLastAccessTime();

            Dictionary<string, string> mmuInstanceMapping = new Dictionary<string, string>();

            //Iterate over each desired MMU
            foreach (string mmuID in mmus)
            {
                MMULoadingProperty mmuLoadingProperty = null;

                //Skip MMU is not contained in adapter
                if (!SessionData.MMULoadingProperties.TryGetValue(mmuID, out mmuLoadingProperty))
                    continue;

                IMotionModelUnitDev mmu = null;

                //Instantiate MMU 
                try
                {
                    mmu = this.mmuInstantiator.InstantiateMMU(mmuLoadingProperty);
                }
                catch (Exception e)
                {
                    Logger.Log(Log_level.L_ERROR, $"Problem at loading MMU {mmuLoadingProperty.Description.Name}, Exception: {e.Message}, {e.StackTrace}");

                    return new Dictionary<string, string>();
                }

                //Assign the service access
                mmu.ServiceAccess = sessionContent.ServiceAccess;

                //Assign the scene
                mmu.SceneAccess = sessionContent.SceneBuffer;

                //Assign a new instance of the skeleton access
                mmu.SkeletonAccess = this.skeletonAccess;//new SkeletonAccess();

                //Set the instance as the adapter
                mmu.AdapterEndpoint = new AdapterEndpoint()
                {
                    Instance = SessionData.AdapterInstance,
                    Description = SessionData.AdapterDescription,
                    MMIRegisterAddress = SessionData.MMIRegisterAddress
                };


                Logger.Log(Log_level.L_INFO, $"Loaded MMU: {mmuLoadingProperty.Description.Name} for session: {sessionID}");

                //Add to the specific avatar content
                AvatarContent avatarContent = null;

                if (!sessionContent.AvatarContent.TryGetValue(idContainer.AvatarID, out avatarContent))
                {
                    avatarContent = new AvatarContent(idContainer.AvatarID);

                    sessionContent.AvatarContent.TryAdd(idContainer.AvatarID, avatarContent);
                }

                //Add the mmu
                avatarContent.MMUs.Add(mmuLoadingProperty.Description.ID, mmu);

                //To do -> create a unique instance ID
                mmuInstanceMapping.Add(mmuLoadingProperty.Description.ID, "tbd");
            }

            return mmuInstanceMapping;
        }


        /// <summary>
        /// Should work
        /// </summary>
        /// <param name="sceneManipulations"></param>
        /// <param name="sessionID"></param>
        public virtual MBoolResponse PushScene(MSceneUpdate sceneUpdates, string sessionID)
        {
            SessionContent sessionContent = null;

            //Get the session content for the id
            MBoolResponse sessionResult = SessionData.GetSessionContent(sessionID, out sessionContent);

            //Skip if invalid session result
            if (!sessionResult.Successful)
            {
                Debug.Fail(sessionResult.LogData.ToString());
                return sessionResult;
            }


            //Set the last access time
            sessionContent.UpdateLastAccessTime();

            //Synchronize the respective scene
            return sessionContent.SceneBuffer.Apply(sceneUpdates);

        }


        /// <summary>
        /// Creates a new session for the specified id
        /// </summary>
        /// <param name="sessionID"></param>
        public virtual MBoolResponse CreateSession(string sessionID)
        {
            //Skip if the sessionID is invalid
            if (sessionID == null || sessionID.Count() == 0)
                return new MBoolResponse(false)
                {
                    LogData = new List<string>() { "Session ID invalid" }
                };

            //Skip if sessionID already available
            if (SessionData.SessionIDAvailable(sessionID))
                return new MBoolResponse(false)
                {
                    LogData = new List<string>() { "Session ID already available" }
                };

            //Create a new session content using the specified id
            SessionContent sessionContent = SessionData.CreateSessionContent(sessionID);

            //Set the last access time
            sessionContent.UpdateLastAccessTime();

            return new MBoolResponse(true);
        }


        /// <summary>
        /// Clsoes the session
        /// </summary>
        /// <param name="sessionID"></param>
        /// <returns></returns>
        public virtual MBoolResponse CloseSession(string sessionID)
        {
            Logger.Log(Log_level.L_INFO, $"Closing the session: {sessionID}");
            return SessionData.RemoveSessionContent(sessionID);
        }



        #endregion

        #region MMU functionality

        /// <summary>
        /// Basic initialization of a MMMU
        /// </summary>
        /// <param name="mmuID"></param>
        /// <param name="sessionID"></param>
        public virtual MBoolResponse Initialize(MAvatarDescription avatarDescription, Dictionary<string, string> properties, string mmuID, string sessionID)
        {
            SessionContent sessionContent = null;
            AvatarContent avatarContent = null;

            MBoolResponse sessionResult = SessionData.GetContents(sessionID, out sessionContent, out avatarContent);
            this.skeletonAccess = new IntermediateSkeleton();
            this.skeletonAccess.InitializeAnthropometry(avatarDescription);

            //Skip if invalid session result
            if (!sessionResult.Successful)
                return sessionResult;

            try
            {
                //Update the access time
                sessionContent.UpdateLastAccessTime();

                //Get the corresponding MMU
                IMotionModelUnitDev mmu = avatarContent.MMUs[mmuID];

                Logger.Log(Log_level.L_INFO, "MMU initialized: " + mmu.Name + " " + sessionID);

                //Call the respective MMU
                return avatarContent.MMUs[mmuID].Initialize(avatarDescription, properties);
            }
            catch (Exception e)
            {
                Logger.Log(Log_level.L_ERROR, $"Problem at initializing MMU: {mmuID}, message: {e.Message}");

                return new MBoolResponse(false)
                {
                    LogData = new List<string>()
                     {
                         e.Message,
                         e.StackTrace,
                         e.InnerException.ToString(),
                         e.StackTrace
                     }
                };

            }
        }


        /// <summary>
        /// Execute command of a MMU
        /// </summary>
        /// <param name="instruction"></param>
        /// <param name="simulationState"></param>
        /// <param name="hierarchy"></param>
        /// <param name="mmuID"></param>
        /// <param name="sessionID"></param>
        public virtual MBoolResponse AssignInstruction(MInstruction instruction, MSimulationState simulationState, string mmuID, string sessionID)
        {
            SessionContent sessionContent = null;
            AvatarContent avatarContent = null;

            MBoolResponse sessionResult = SessionData.GetContents(sessionID, out sessionContent, out avatarContent);

            //Directly return if not successfull
            if (!sessionResult.Successful)
                return sessionResult;

            sessionContent.UpdateLastAccessTime();


            Logger.Log(Log_level.L_DEBUG, $"Execute instruction {instruction.Name}, {mmuID}");


            //Directly assign the instruction
            return avatarContent.MMUs[mmuID].AssignInstruction(instruction, simulationState);        
        }


        /// <summary>
        /// Basic do step routine which triggers the simulation update of the repsective MMU
        /// </summary>
        /// <param name="time"></param>
        /// <param name="simulationState"></param>
        /// <param name="mmuID"></param>
        /// <param name="sessionID"></param>
        /// <returns></returns>
        public virtual MSimulationResult DoStep(double time, MSimulationState simulationState, string mmuID, string sessionID)
        {
            SessionContent sessionContent = null;
            AvatarContent avatarContent = null;

            MBoolResponse sessionResult = SessionData.GetContents(sessionID, out sessionContent, out avatarContent);

            //Skip if invalid session result
            if (!sessionResult.Successful)
                return null;

            sessionContent.UpdateLastAccessTime();

            //Execute the do step of the respective MMU
            return avatarContent.MMUs[mmuID].DoStep(time, simulationState);
        }


        public virtual MBoolResponse CheckPrerequisites(MInstruction instruction, string mmuID, string sessionID)
        {
            SessionContent sessionContent = null;
            AvatarContent avatarContent = null;

            MBoolResponse sessionResult = SessionData.GetContents(sessionID, out sessionContent, out avatarContent);

            if (!sessionResult.Successful)
                return sessionResult;


            sessionContent.UpdateLastAccessTime();

            //Execute the method of the MMU
            return avatarContent.MMUs[mmuID].CheckPrerequisites(instruction);
        }


        public virtual MBoolResponse Abort(string instructionId, string mmuID, string sessionID)
        {
            SessionContent sessionContent = null;
            AvatarContent avatarContent = null;

            MBoolResponse sessionResult = SessionData.GetContents(sessionID, out sessionContent, out avatarContent);

            if (!sessionResult.Successful)
                return sessionResult;

            sessionContent.UpdateLastAccessTime();

            //Abort the respective MMU
            return avatarContent.MMUs[mmuID].Abort(instructionId);
        }


        public virtual MMUDescription GetDescription(string mmuID, string sessionID)
        {
            return SessionData.MMUDescriptions.Find(s => s.Name == mmuID);
        }


        public virtual List<MConstraint> GetBoundaryConstraints(MInstruction instruction, string mmuID, string sessionID)
        {
            SessionContent sessionContent = null;
            AvatarContent avatarContent = null;

            MBoolResponse sessionResult = SessionData.GetContents(sessionID, out sessionContent, out avatarContent);

            if (!sessionResult.Successful)
                return new List<MConstraint>();

            sessionContent.UpdateLastAccessTime();

            return avatarContent.MMUs[mmuID].GetBoundaryConstraints(instruction);
        }

        public virtual MBoolResponse Dispose(string mmuID, string sessionID)
        {
            SessionContent sessionContent = null;
            AvatarContent avatarContent = null;

            MBoolResponse sessionResult = SessionData.GetContents(sessionID, out sessionContent, out avatarContent);

            if (!sessionResult.Successful)
                return sessionResult;

            sessionContent.UpdateLastAccessTime();

            //Call the dispose method of the respective MMU
            return avatarContent.MMUs[mmuID].Dispose(new Dictionary<string, string>());
        }

        public virtual List<MMUDescription> GetMMus(string sessionID)
        {
            SessionContent sessionContent = null;
            AvatarContent avatarContent = null;

            MBoolResponse sessionResult = SessionData.GetContents(sessionID, out sessionContent, out avatarContent);

            if (!sessionResult.Successful)
                return new List<MMUDescription>();

            sessionContent.UpdateLastAccessTime();

            return SessionData.MMUDescriptions.Where(s => avatarContent.MMUs.Keys.Contains(s.ID)).ToList();
        }


        /// <summary>
        /// Creates a checkpoint for all specified MMUs.
        /// The checkpoint contains the internal state of each MMU whoch can be later used to restore the state.
        /// </summary>
        /// <param name="mmuIDs"></param>
        /// <param name="sessionID"></param>
        /// <param name="checkpointID"></param>
        public virtual byte[] CreateCheckpoint(string mmuID, string sessionID)
        {
            SessionContent sessionContent = null;
            AvatarContent avatarContent = null;

            MBoolResponse sessionResult = SessionData.GetContents(sessionID, out sessionContent, out avatarContent);

            if (!sessionResult.Successful)
                return null;

            sessionContent.UpdateLastAccessTime();

            //Add method to interface
            byte[] checkpointData = avatarContent.MMUs[mmuID].CreateCheckpoint();

            Logger.Log(Log_level.L_INFO, $"Checkpoint of {mmuID} sucessfully created ({checkpointData.Length} bytes)");

            return checkpointData;
        }

        public virtual MBoolResponse RestoreCheckpoint(string mmuID, string sessionID, byte[] checkpointData)
        {
            SessionContent sessionContent = null;
            AvatarContent avatarContent = null;

            MBoolResponse sessionResult = SessionData.GetContents(sessionID, out sessionContent, out avatarContent);

            if (!sessionResult.Successful)
                return sessionResult;

            sessionContent.UpdateLastAccessTime();

            Logger.Log(Log_level.L_INFO, $"Restore checkpoint of {mmuID}");


            return avatarContent.MMUs[mmuID].RestoreCheckpoint(checkpointData);
        }

        public Dictionary<string, string> ExecuteFunction(string name, Dictionary<string, string> parameters, string mmuID, string sessionID)
        {
            SessionContent sessionContent = null;
            AvatarContent avatarContent = null;

            MBoolResponse sessionResult = SessionData.GetContents(sessionID, out sessionContent, out avatarContent);

            if (!sessionResult.Successful)
                return null;

            sessionContent.UpdateLastAccessTime();


            Logger.Log(Log_level.L_DEBUG, $"Ecexute function {name} of {mmuID}");

            return avatarContent.MMUs[mmuID].ExecuteFunction(name, parameters);

        }

        /// <summary>
        /// Returns the assigned adapter description
        /// </summary>
        /// <returns></returns>
        public virtual MAdapterDescription GetAdapterDescription()
        {
            return SessionData.AdapterDescription;
        }








        #endregion
    }
}

File: Framework/LanguageSupport/python/MMIPython/src/MMIPython/adapter/thrift_adapter_implementation.py
## SPDX-License-Identifier: MIT
## The content of this file has been developed in the context of the MOSIM research project.
## Original author(s): Jannes Lehwald, Janis Sprenger


 #-*- coding: utf-8 -*-
"""

"""
import os
import json
from datetime import datetime
from MMIStandard.avatar.ttypes import MAvatarDescription, MAvatarDescription
from MMIStandard.scene.ttypes import MSceneUpdate, MSceneManipulation
from MMIStandard.core.ttypes import MBoolResponse, MIPAddress
from MMIStandard.mmu.ttypes import MSimulationState, MInstruction
from MMIStandard.register.ttypes import MAdapterDescription
from MMIStandard.register import MMIAdapter, MMIRegisterService
#from MMIStandard.ttypes import MAvatarPosture, MInstruction, MAvatarState, MAvatarDescription, MSceneManipulation, MIPAddress
#from MMIStandard import MMIAdapter

from MMIPython.core.utils.description_utils import create_mmu_description
from MMIPython.core.services.service_access import ServiceAccess

from MMIPython.adapter.session.session_data import SessionData

from thrift.transport import TSocket
from thrift.transport import TTransport
from thrift.protocol import TCompactProtocol
from thrift.server import TServer

import MMIPython.core.scene.remote_scene_access as scene_access

class ThriftAdapterImplementation(MMIAdapter.Iface):
    
    """
    The implementation of the MMIAdapter
    
    """
    
    def __init__(self, session_data): #r_ip_address):

        """
        The basic constructor of the adapter
        
        Parameters
        ----------
        r_ip_address : MIPAddress
            The register MIPAddress used for the services
        """
        
        #assert (isinstance(r_ip_address, MIPAddress)),"r_ip_address is no MIPAddress"
        
        self.session_data = session_data       
   
 
        if __debug__:
            print('Adapter started in debug mode')
        else:
            print('Adapter started in optimized mode')
        
        
       
    
    def Initialize(self, avatarDescription, properties, mmuID, sessionID):
        """
        Basic initialization of a MMU
        
        Parameters
        ----------
        referenceAvatar : MAvatarDescription
            The reference avatar description
        properties : dict
            The properties for initialization of the MMU
        mmuID : str
            The mmu ID
        sessionID : str
            The session ID
            
        Returns
        -------
        MBoolResponse
        """
        if __debug__:
            print('--------------')
        
        assert (isinstance(sessionID, str)),"SessionID is no string"
        assert (isinstance(mmuID, str)),"mmuID is no string"
        assert (isinstance(avatarDescription, MAvatarDescription)),"avatarDescritption is no MAvatarDescription"
        assert (isinstance(properties, dict)),"properties is no dict"
        assert (all(isinstance(x, str) for x in properties.keys())), "Not all keys are of type string in properties"
        assert (all(isinstance(x, str) for x in properties.values())), "Not all values are of type string in properties"

        
        # get the mmu
        mmu = self.session_data.GetMMU(sessionID, mmuID)
        if mmu is not None:
            self.session_data.last_access =datetime.now()
            return mmu.Initialize(avatarDescription, properties)
    
        else:
             return MBoolResponse(False,["MMU " + mmuID + "with session :" + sessionID+  " could not be initialized"])

        """
        if mmu is not None:
            result = mmu.initialize(avatarDescription, properties)
        
            # Update last access
            SessionData.last_access =datetime.now()
            
            if initialized:
                print('Initialized MMU {1} with session : {0}'.format(sessionID, mmuID))
            else:
                print('MMU {1} with session : {0} could not be initialized'.format(sessionID, mmuID))
            
        else:
            print('MMU {1} with session : {0} not found'.format(sessionID, mmuID))
        
        if __debug__:
            print('--------------')
        """
        #return initialized
        

    def AssignInstruction(self, instruction, simulationState, mmuID, sessionID):
        """
        Assigns a command to a MMU
        
        Parameters
        ----------
        instruction : MInstruction
            The reference avatar posture
        avatarState : MAvatarState
            The MAvatarState
        mmuID : str
            The mmu ID
        sessionID : str
            The session ID
            
        Returns
        -------
        MBoolResponse
        """
        
        assert (isinstance(sessionID, str)),"SessionID is no string"
        assert (isinstance(mmuID, str)),"mmuID is no string"
        assert (isinstance(instruction, MInstruction)), "instruction is no MInstruction"
        assert (isinstance(simulationState, MSimulationState)), "simulationState is no MSimulationState"
        
        print('Assign instruction to MMU {1} with session : {0}'.format(sessionID, mmuID))
        
        
        # Get the avatar content for the stored posture hierarchy
        avatar_content = self.session_data.get_avatar_content(sessionID)
        
        # Update last access
        self.session_data.last_access =datetime.now()
        
        # get mmu
        mmu = self.session_data.GetMMU(sessionID, mmuID)
        
        if mmu is not None:
            # Assign instruction
            return mmu.AssignInstruction(instruction, simulationState)
        else:
            return MBoolResponse(False,["MMU " + mmuID + "with session :" + sessionID+  " could not be found"])
      

    def DoStep(self, time, simulationState, mmuID, sessionID):
        """
        Basic do step routine which triggers the simulation update of the repsective MMU
        
        Parameters
        ------------
        time : float
            The simulation time
        avatarState : MAvatarState
            The incoming MAvatarState
        mmuID : str
            The mmuID
        sessionID : str
            The session ID
        
        Returns
        -------
        MMUResult
            The result of the MMU
        """
        
        assert (isinstance(sessionID, str)),"SessionID is no string"
        assert (isinstance(mmuID, str)),"mmuID is no string"
        assert (isinstance(simulationState, MSimulationState)), "simulationState is no MSimulationState"
        assert (isinstance(time, float)),"time is no float"
        
        mmu_result = None
        print("Do Step")
        
        # Get the avatar content for the stored posture hierarchy
        avatar_content = self.session_data.get_avatar_content(sessionID)
        
        # get mmu
        mmu = self.session_data.GetMMU(sessionID, mmuID)
        # Update last access
        self.session_data.last_access =datetime.now()

        if mmu is not None:
            # Do step
           return mmu.DoStep(time,simulationState)
        else:
           print('MMU {1} with session : {0} could net be found'.format(sessionID, mmuID))
           return None

    def GetBoundaryConstraints(self, instruction, mmuID, sessionID):
        """
        Returns constraints which are relevant for the transition
        
        Parameters
        ------------
        instruction : MInstruction
            The given motion instruction
        mmuID : str
            The mmuID
        sessionID : str
            The session ID
        
        Returns
        -------
        list<MConstraint>
        """

        assert (isinstance(sessionID, str)),"SessionID is no string"
        assert (isinstance(mmuID, str)),"mmuID is no string"
        assert (isinstance(instruction, MInstruction)), "instruction is no MInstruction"

        print('Get BoundardyConstraints of the MMU {1} with session : {0}'.format(sessionID, mmuID))

          # get mmu
        mmu = self.session_data.GetMMU(sessionID, mmuID)
        
        if mmu is not None:
            return mmu.GetBoundaryConstraints(instruction)
        else:
            print('MMU {1} with session : {0} could net be found'.format(sessionID, mmuID))
            return None
          
           

    def CheckPrerequisites(self, instruction, mmuID, sessionID):
        """
        
        Returns the prerequisites to start the given motion instruction
        
        Parameters
        -------------
        instruction : MInstruction
            The given motion instruction
        mmuID : str
            The mmuID
        sessionID : str
            The session ID
            
        Returns
        ----------
        list<MConstraints>
            Prerequisites
        """
        
        assert (isinstance(sessionID, str)),"SessionID is no string"
        assert (isinstance(mmuID, str)),"mmuID is no string"
        assert (isinstance(instruction, MInstruction)), "instruction is no MInstruction"
        
        print('Check prerequisites of the MMU {1} with session : {0}'.format(sessionID, mmuID))
        
        
        # get mmu
        mmu = self.session_data.GetMMU(sessionID, mmuID)
        
        if mmu is not None:
            
#            if __debug__:
#                if mmu.name == "PythonAbstractionTestMMU":
#                    
#                    sessID, avID = session_utils.get_splitted_IDs(sessionID)
#                    
#                    if instruction.Properties is None:
#                        instruction.Properties = dict()
#                        
#                    instruction.Properties["sessionID"] = sessID
#                    instruction.Properties["avatarID"] = avID
#                    instruction.Properties["manageScene"] = "False"
#                    instruction.Properties["address"] = self._session_data.service_access._register_ip_address
                
            # Get prerequisites
            return mmu.CheckPrerequisites(instruction)
        else:
            print('MMU {1} with session : {0} could net be found'.format(sessionID, mmuID))
            return None
    
    def CloseSession(self, sessionID):
        return self.session_data.removeSessionContent(sessionID)

    def Abort(self, instructionID, mmuID, sessionID):
        """
        
        Forces the termination of a MMU
        
        Parameters
        ------------
        mmuID : str
            The mmuID
        sessionID : str
            The session ID
            
        Returns
        -------
        MBoolResponse
        """
        
        assert (isinstance(sessionID, str)),"SessionID is no string"
        assert (isinstance(mmuID, str)),"mmuID is no string"
        
        print('Abort the MMU {1} with session : {0}'.format(sessionID, mmuID))
        # Update last access
        self.session_data.last_access =datetime.now()
        
        # get mmu
        mmu = self.session_data.GetMMU(sessionID, mmuID)
        
        if mmu is not None:
            # Abort mmu
            return mmu.Abort(instructionID)
        else:
            return MBoolResponse(False,["MMU " + mmuID + "with session :" + sessionID+  " could not be found"])



    def Dispose(self, mmuID, sessionID):
        """
        
        Called to release ressources from a MMU.
        
        Parameters
        ------------
        mmuID : str
            The mmuID
        sessionID : str
            The session ID
            
        Returns
        -------
        MBoolResponse
        """
        
        assert (isinstance(sessionID, str)),"SessionID is no string"
        assert (isinstance(mmuID, str)),"mmuID is no string"
        
 
        
        print('Dispose the MMU {1} with session : {0}'.format(sessionID, mmuID))
        
        # get mmu
        mmu = self.session_data.GetMMU(sessionID, mmuID)
        
        if mmu is not None:
            # Dispose mmu
            return mmu.Dispose()
        else:
            return MBoolResponse(False,["MMU " + mmuID + "with session :" + sessionID+  " could not be found"])  

    def GetStatus(self):
        """
        Returns the status of this adapter as dictionary.
        
        Returns
        ---------
        dict
            The status as key value pairs.
        """
        
        #print('Get the adapter status')
        
        status = dict()

        status["Version"] = '0.0.2'
        status["Running since"] = self.session_data.start_time.strftime("%Y-%m-%d %H:%M:%S")
        status["Total Sessions"] = str(len(self.session_data.session_contents))
        status["Loadable MMUs"] = str(len(self.session_data.mmu_descriptions))

        status["Version"] = "5";

        if self.session_data.last_access is None:
            status["Last Access"] = "None"
        else:
            status["Last Access"] = self.session_data.last_access.strftime("%Y-%m-%d %H:%M:%S")
        
        
        return status;

    def GetAdapterDescription(self):
        return self.session_data.adapter_description

    def CreateSession(self, sessionID):
        """
        
        Creates a session with a given session ID
        
        Parameters
        ----------
        sessionID : str
            The session ID
       
        Returns
        -------
        MBoolResponse
        """
        
        assert (isinstance(sessionID, str)),"SessionID is no string"
        
        print('Create a session with ID : {0}'.format(sessionID))
        
        # Create session content
        self.session_data.create_session_content(sessionID)
        
        return MBoolResponse(True,None)

    def PushScene(self, sceneUpdates, sessionID):
        """
        Method to synchronize the scene
        
        Parameters
        ----------
        sceneUpdates: MSceneUpdate
            The scene changes
        sessionID : str
            The session ID
       
        Returns
        -------
        MBoolResponse
        """
        assert (isinstance(sessionID, str)),"SessionID is no string"
        assert (isinstance(sceneUpdates, MSceneUpdate)),"sceneUpdates is no MSceneUpdate"
    
        # Get the session content
        session_content = self.session_data.get_session_content(sessionID)
        
        # if session_content is not None:
        #     return session_content.scene_access.apply(sceneUpdates)

        return MBoolResponse(True)



   

    def GetLoadableMMUs(self):
        """
        Returns the description of all loadable MMUs from this adapter.
        
        Returns
        ---------
        list<MMUDescription>
            The loadable MMUs.
        """
        
        #print('Collect loadable MMUs')
        return self.session_data.mmu_descriptions

    def GetMMus(self, sessionID):
        """
        Returns all loaded MMUs for the given session
        
        Parameters
        -----------
        sessionID : str
            The session ID
        """
        assert (isinstance(sessionID, str)),"SessionID is no string"
        
        print('Get loaded MMUs for session : {0}'.format(sessionID))
        
        mmus = list()
        
        mmus = self.session_data.GetLoadedMMUs(sessionID)
        # # Get the session content
        # avatar_content = self.session_data.get_avatar_content(sessionID, True)
        
        # if avatar_content is not None:
            
        #     # Get the mmus
        #     mmus = list(avatar_content.mmus.keys())
        mmus = [self.GetDescription(x, sessionID) for x in mmus]
            
        print('Return loaded MMUs {1} for session : {0}'.format(sessionID, mmus))
            
        return mmus

    def GetDescription(self, mmuID, sessionID):
        """
        
        Returns the description of a MMU.
        
        Parameters
        ------------
        mmuID : str
            The mmuID
        sessionID : str
            The session ID
        """
        
        print('Get description for session : {0}'.format(sessionID))
        
        return next((x for x in self.session_data.mmu_descriptions if x.ID == mmuID), None)

    def GetScene(self, sessionID):
        """
        Returns all objects from the scene with the given session ID.
        
        Returns
        ---------
        list<MSceneObject>
            The scene objects.
        """
        
        assert (isinstance(sessionID, str)),"SessionID is no string"
        
        print('Get scene for session : {0}'.format(sessionID))
        
        
        # get session content
        session_content = self.session_data.get_session_content(sessionID)
        
        if session_content is not None:
            
            # get all scene obects from the scene access
            return session_content.scene_access.GetSceneObjects()
            
        

    def GetSceneChanges(self, sessionID):
        """
        Returns all scene manipulations from the scene with the given session ID.
        
        Returns
        ---------
        list<MSceneManipulation>
            The scene manipulations.
        """
        
        assert (isinstance(sessionID, str)),"SessionID is no string"
        
        print('Get scene changes for session : {0}'.format(sessionID))
        
        scene_manipulations = list()
        
        # get session content
        session_content = self.session_data.get_session_content(sessionID)
        
        if session_content is not None:
            
            # get all scene obects from the scene access
            scene_manipulations = session_content.scene_access.scene_events
            
        return scene_manipulations

    def LoadMMUs(self, mmus, sessionID):
        """
       
        Loads all selected mmus to the given session.
        
        Parameters
        ------------
        mmus : list
            The mmu names
        sessionID : str
            The session ID
            
        Returns
        -------
        MBoolResponse
        """
        assert (isinstance(sessionID, str)),"SessionID is no string"
        assert (isinstance(mmus, list)),"mmus is no list"
        
        print('Load mmus for session : {0}'.format(sessionID))

        print("mmus: ", mmus)
        check = {}

        for mmu in mmus:
            if self.session_data.LoadMMU(sessionID, mmu):
                check[mmu] = "tbd"
        
        return check
                    

    def CreateCheckpoint(self, mmuID, sessionID):
        """
        Creates a checkpoint for the specified MMU.
        The checkpoint contains the internal state of each MMU which can be later used to restore the state.
        
        Parameters:
        ----------
        mmuID : str
            The mmu ID
        sessionID : str
            The session ID
        """
        
        assert (isinstance(sessionID, str)),"SessionID is no string"
        assert (isinstance(mmuID, str)),"mmuID is no str"
        
        print('Create checkpoint for mmu : {1} with session : {0}'.format(sessionID, mmuID))

          
        # get mmu
        mmu = self.session_data.GetMMU(sessionID, mmuID)

        if mmu is not None:
           return mmu.CreateCheckPoint() 
        else:
            print('MMU {1} with session : {0} could net be found'.format(sessionID, mmuID))
            return None

        
        
    def RestoreCheckpoint(self, mmuID, sessionID, data):
        """
        Restores a checkpoint for the specified MMU.
        
        Parameters:
        ----------
        mmuID : str
            The mmu IDs
        sessionID : str
            The session ID
        data : bytearray
            The checkpoint data
            
        Returns
        -------
        bool
        """
        
        assert (isinstance(sessionID, str)),"SessionID is no string"
        assert (isinstance(mmuID, str)),"mmuID is no string"
        print(type(data))
        assert (isinstance(data, bytes)),"data is no bytes"
        
        print('Restore checkpoint for mmu {1} with session : {0} with checkpoint : {1}'.format(sessionID, mmuID))
        
         # get mmu
        mmu = self.session_data.GetMMU(sessionID, mmuID)

        if mmu is not None:
           return mmu.RestoreCheckPoint(data)
        else:
            print('MMU {1} with session : {0} could net be found'.format(sessionID, mmuID))
            return None


    def ExecuteFunction(self, name, parameters, mmuID, sessionID):
        assert (isinstance(name, str)),"SessionID is no string"
        assert (isinstance(parameters, dict)),"SessionID is no string"

        # get mmu
        mmu = self.session_data.GetMMU(sessionID, mmuID)

        if mmu is not None:
           return mmu.ExecuteFunction(name,parameters)
        else:
            print('MMU {1} with session : {0} could net be found'.format(sessionID, mmuID))
            return None

    def create_response(self,result,message):
        if result.LogData is None:
            result.LogData=[message]
        else:
            result.LogData.append(message)
        
        return result



        